### PR TITLE
Extended TSql Parser and Lexer

### PR DESCRIPTION
**Pull request by ealight**
- Added all the math functions to built-ins. ([Source](https://learn.microsoft.com/en-us/sql/t-sql/functions/mathematical-functions-transact-sql?view=sql-server-ver16))
- Extended date built-ins functions up-to-date. ([Source](https://learn.microsoft.com/en-us/sql/t-sql/functions/date-and-time-data-types-and-functions-transact-sql?view=sql-server-ver16))
- Added procedure's temporary tables recognizing. Now temporary tables which start with '#' are going to temp_table_name/full_temp_table_name contexts. ([Source](https://www.red-gate.com/simple-talk/databases/sql-server/t-sql-programming-sql-server/temporary-tables-in-sql-server/))
- Added database engine stored procedures handling. For now it's only 'sp_executesql' but there is a field to easy work with. ([Source](https://learn.microsoft.com/en-us/sql/relational-databases/system-stored-procedures/database-engine-stored-procedures-transact-sql?view=sql-server-ver16)).
- Added old fashion join - NON ANSI Join handling. They're going to own context. ([Source](https://sqlenlight.com/support/help/sa0006/))
- Added dateparts for date built-in functions to match documentation. (dateparts_9 - [DATE_BUCKET](https://learn.microsoft.com/en-us/sql/t-sql/functions/date-bucket-transact-sql?view=sql-server-ver16), dateparts_12 - [DATE_ADD](https://learn.microsoft.com/en-us/sql/t-sql/functions/dateadd-transact-sql?view=sql-server-ver16), [DATE_DIFF](https://learn.microsoft.com/en-us/sql/t-sql/functions/datediff-transact-sql?view=sql-server-ver16), [DATEDIFF_BIG](https://learn.microsoft.com/en-us/sql/t-sql/functions/datediff-big-transact-sql?view=sql-server-ver16), dateparts_15 - [DATENAME](https://learn.microsoft.com/en-us/sql/t-sql/functions/datename-transact-sql?view=sql-server-ver16), [DATEPART](https://learn.microsoft.com/en-us/sql/t-sql/functions/datepart-transact-sql?view=sql-server-ver16), dateparts_datetrunc - [DATETRUNC](https://learn.microsoft.com/en-us/sql/t-sql/functions/datetrunc-transact-sql?view=sql-server-ver16))

### CODE REVIEW

File: sql/tsql/TSqlParser.g4 [hunk 1]
```diff
@@ -4386,17 +4491,41 @@ entity_name_for_parallel_dw
...
+
+persist_table_name
     : (database=id_ '.' (schema=id_)? '.' | schema=id_ '.')? table=id_
     | (database=id_ '.' (schema=id_)? '.' | schema=id_ '.')? blocking_hierarchy=BLOCKING_HIERARCHY
     ;
 
+// https://www.red-gate.com/simple-talk/databases/sql-server/t-sql-programming-sql-server/temporary-tables-in-sql-server/
+temp_table_name
+    : (database=id_ '.' (schema=id_)? '.' | schema=id_ '.')? table=TEMP_ID
+    | (database=id_ '.' (schema=id_)? '.' | schema=id_ '.')? blocking_hierarchy=BLOCKING_HIERARCHY
```
Comment (KvanTTT) [inline sql/tsql/TSqlParser.g4:4526]: Please factor the common part:

```suggestion
    : (database=id_ '.' schema=id_? '.' | schema=id_ '.')? (table=TEMP_ID | blocking_hierarchy=BLOCKING_HIERARCHY)
```

File: sql/tsql/TSqlParser.g4 [hunk 2]
```diff
@@ -4386,17 +4491,41 @@ entity_name_for_parallel_dw
...
     |                       server=id_    '.' database=id_ '.'  schema=id_   '.'
     |                                         database=id_ '.' (schema=id_)? '.'
     |                                                           schema=id_    '.')? table=id_
     ;
 
+// https://www.red-gate.com/simple-talk/databases/sql-server/t-sql-programming-sql-server/temporary-tables-in-sql-server/
+temp_full_table_name
+    : (linkedServer=id_ '.' '.' schema=id_   '.'
+    |                       server=id_    '.' database=id_ '.'  schema=id_   '.'
+    |                                         database=id_ '.' (schema=id_)? '.'
```
Comment (KvanTTT) [inline sql/tsql/TSqlParser.g4:4509]: Redudant parentheses:

```suggestion
    |                                         database=id_ '.' schema=id_? '.'
```

**Review by KvanTTT — CHANGES_REQUESTED** [review]
Please cover your changes with some examples (`sql/tsql/examples`).

**Comment by ealight** [comment]
@KvanTTT 

- Added dateparts for date built-in functions to match documentation. (dateparts_9 - [DATE_BUCKET](https://learn.microsoft.com/en-us/sql/t-sql/functions/date-bucket-transact-sql?view=sql-server-ver16), dateparts_12 - [DATE_ADD](https://learn.microsoft.com/en-us/sql/t-sql/functions/dateadd-transact-sql?view=sql-server-ver16), [DATE_DIFF](https://learn.microsoft.com/en-us/sql/t-sql/functions/datediff-transact-sql?view=sql-server-ver16), [DATEDIFF_BIG](https://learn.microsoft.com/en-us/sql/t-sql/functions/datediff-big-transact-sql?view=sql-server-ver16), dateparts_15 - [DATENAME](https://learn.microsoft.com/en-us/sql/t-sql/functions/datename-transact-sql?view=sql-server-ver16), [DATEPART](https://learn.microsoft.com/en-us/sql/t-sql/functions/datepart-transact-sql?view=sql-server-ver16), dateparts_datetrunc - [DATETRUNC](https://learn.microsoft.com/en-us/sql/t-sql/functions/datetrunc-transact-sql?view=sql-server-ver16))
- Added examples for each case.
- Fixed 'hints.sql.tree' to test tree correctly with new changes.

**Comment by ealight** [comment]
- Fixed '[xml_data_type.sql.tree](https://github.com/antlr/grammars-v4/pull/2994/files#diff-e09bd0b5a704c4cc2c6ebcb68641a5c283856b6ded48f54e75038bcd2f5ad28b)' to test tree correctly with new changes.

File: sql/tsql/TSqlParser.g4 [hunk 3]
```diff
@@ -4115,6 +4220,43 @@ xml_data_type_methods
...
+dateparts_9
+    : (YEAR | YEAR_ABBR)
+    | (QUARTER | QUARTER_ABBR)
+    | (MONTH | MONTH_ABBR)
+    | (DAY | DAY_ABBR)
+    | (WEEK | WEEK_ABBR)
+    | (HOUR | HOUR_ABBR)
+    | (MINUTE | MINUTE_ABBR)
+    | (SECOND | SECOND_ABBR)
+    | (MILLISECOND | MILLISECOND_ABBR)
```
Comment (KvanTTT) [inline sql/tsql/TSqlParser.g4:4233]: Parentheses are excess:

```suggestion
    : YEAR | YEAR_ABBR
    | QUARTER | QUARTER_ABBR
    | MONTH | MONTH_ABBR
    | DAY | DAY_ABBR
    | WEEK | WEEK_ABBR
    | HOUR | HOUR_ABBR
    | MINUTE | MINUTE_ABBR
    | SECOND | SECOND_ABBR
    | MILLISECOND | MILLISECOND_ABBR
```

File: sql/tsql/TSqlParser.g4 [hunk 4]
```diff
@@ -4115,6 +4220,43 @@ xml_data_type_methods
...
+    ;
+
+// https://learn.microsoft.com/en-us/sql/t-sql/functions/dateadd-transact-sql?view=sql-server-ver16
+dateparts_12
+    :
+    | dateparts_9
+    | (DAYOFYEAR | DAYOFYEAR_ABBR)
+    | (MICROSECOND | MICROSECOND_ABBR)
+    | (NANOSECOND | NANOSECOND_ABBR)
+    ;
```
Comment (KvanTTT) [inline sql/tsql/TSqlParser.g4:4243]: Is it ok that `dateparts_12` has the empty alternative? If so, plese remove the empty alternative here and insert `?` to all `dateparts_12` usings. Also, parentheses are excess.

```suggestion
dateparts_12
    :
    | dateparts_9
    | DAYOFYEAR | DAYOFYEAR_ABBR
    | MICROSECOND | MICROSECOND_ABBR
    | NANOSECOND | NANOSECOND_ABBR
    ;
```

File: sql/tsql/TSqlParser.g4 [hunk 5]
```diff
@@ -4115,6 +4220,43 @@ xml_data_type_methods
...
+
+// https://learn.microsoft.com/en-us/sql/t-sql/functions/datename-transact-sql?view=sql-server-ver16
+dateparts_15
+    : dateparts_12
+    | (WEEKDAY | WEEKDAY_ABBR)
+    | (TZOFFSET | TZOFFSET_ABBR)
+    | (ISO_WEEK | ISO_WEEK_ABBR)
+    ;
```
Comment (KvanTTT) [inline sql/tsql/TSqlParser.g4:4251]:
```suggestion
dateparts_15
    : dateparts_12
    | WEEKDAY | WEEKDAY_ABBR
    | TZOFFSET | TZOFFSET_ABBR
    | ISO_WEEK | ISO_WEEK_ABBR
    ;
```

File: sql/tsql/TSqlParser.g4 [hunk 6]
```diff
@@ -4115,6 +4220,43 @@ xml_data_type_methods
...
+
+// https://learn.microsoft.com/en-us/sql/t-sql/functions/datetrunc-transact-sql?view=sql-server-ver16
+dateparts_datetrunc
+    : dateparts_9
+    | (DAYOFYEAR | DAYOFYEAR_ABBR)
+    | (MICROSECOND | MICROSECOND_ABBR)
+    | (ISO_WEEK | ISO_WEEK_ABBR);
```
Comment (KvanTTT) [inline sql/tsql/TSqlParser.g4:4258]: Excess parentheses + formatting (semicolon should be on new line):

```suggestion
dateparts_datetrunc
    : dateparts_9
    | DAYOFYEAR | DAYOFYEAR_ABBR
    | MICROSECOND | MICROSECOND_ABBR
    | ISO_WEEK | ISO_WEEK_ABBR
    ;
```

File: sql/tsql/TSqlParser.g4 [hunk 7]
```diff
@@ -4386,15 +4528,37 @@ entity_name_for_parallel_dw
...
+
 table_name
-    : (database=id_ '.' (schema=id_)? '.' | schema=id_ '.')? table=id_
-    | (database=id_ '.' (schema=id_)? '.' | schema=id_ '.')? blocking_hierarchy=BLOCKING_HIERARCHY
+    : temp_table_name
+    | persist_table_name
+    ;
+
+persist_table_name
+    : (database=id_ '.' (schema=id_)? '.' | schema=id_ '.')? (table=id_ | blocking_hierarchy=BLOCKING_HIERARCHY)
```
Comment (KvanTTT) [inline sql/tsql/TSqlParser.g4:4556]:
```suggestion
    : (database=id_ '.' schema=id_? '.' | schema=id_ '.')? (table=id_ | blocking_hierarchy=BLOCKING_HIERARCHY)
```

File: sql/tsql/TSqlParser.g4 [hunk 8]
```diff
@@ -4386,15 +4528,37 @@ entity_name_for_parallel_dw
...
+    ;
+
+// https://www.red-gate.com/simple-talk/databases/sql-server/t-sql-programming-sql-server/temporary-tables-in-sql-server/
+temp_table_name
+    : (database=id_ '.' (schema=id_)? '.' | schema=id_ '.')? (table=TEMP_ID | blocking_hierarchy=BLOCKING_HIERARCHY)
```
Comment (KvanTTT) [inline sql/tsql/TSqlParser.g4:4561]:
```suggestion
    : (database=id_ '.' schema=id_? '.' | schema=id_ '.')? (table=TEMP_ID | blocking_hierarchy=BLOCKING_HIERARCHY)
```

**Review by KvanTTT — CHANGES_REQUESTED** [review]

**Comment by KvanTTT** [comment]
Welcome to the grammars repository and thanks for the contribution. But please fix remaining issues before merging.

File: sql/tsql/TSqlParser.g4 (hunk 3 above)
Comment (ealight) [reply sql/tsql/TSqlParser.g4]: Sure, I wanted to highlight datepart with their abbreviations but to lightweight parser it's really worth to remove the parentheses.

File: sql/tsql/TSqlParser.g4 (hunk 4 above)
Comment (ealight) [reply sql/tsql/TSqlParser.g4]: Oh, empty alternative isn't okey here. This is my inattention. Parentheses also deleted.

**Comment by ealight** [comment]
@KvanTTT, Done.

**Comment by pguyot** [comment]
@ealight thank you for your contribution.

I am confused by the temp table rule changes. It makes rules `full_table_name` and `table_name` ambiguous which slows down parsing (about 5-10% on a quick benchmark). Indeed, `id_` can be a `TEMP_ID`.

Should we revert this?

**Comment by ealight** [comment]
@pguyot hi, it would be great to leave distinction between temporary tables and persisted tables (according to documentation). If you have better solution without slowing down the parsing, you are welcome to revert current changes. Thanks.

**Comment by pguyot** [comment]
My experience is that `temp_full_table_name` and `persist_full_table_name` are pretty useless because beyond ambiguity, temporary table names can be accepted by `persist_full_table_name` if they are quoted with brackets `[#table]`.

I'm not sure grammar can catch this and you need to parse the identifier eventually to distinguish temporary tables and persistent tables.

So I think we should revert this change.

**Pull request merged**